atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java

```diff
@@ -351,12 +351,10 @@ public Optional<CassandraClientPoolingContainer> getRandomGoodHostForPredicate(
                 .keys()
                 .collect(Collectors.toSet());
 
-        Set<CassandraServer> hostsInPermittedDatacenters = hostsMatchingPredicate.stream()
-                .filter(pool -> {
-                    String datacenter = hostToDatacenter.get(pool);
-                    return datacenter == null || !maximallyAttemptedDatacenters.contains(datacenter);
-                })
-                .collect(Collectors.toSet());
+        Set<CassandraServer> hostsInPermittedDatacenters = Sets.filter(hostsMatchingPredicate, pool -> {
+            String datacenter = hostToDatacenter.get(pool);
+            return datacenter == null || !maximallyAttemptedDatacenters.contains(datacenter);
+        });
```

Comment:
As mentioned before: the downside is that if we use the old stream, we can collect to `ImmutableSet` and thus pass that in to the blacklist to guarantee its immutability.